Remove ConnectionString on With methods in Hosting App Model

eerhardt opened the pull request:
Instead, add new Add overloads that take a connection string to an existing server.

Note - I needed to make `IDistributedApplicationComponentBuilder<out T>` an `out` generic because I need `IDistributedApplicationComponentBuilder<PostgresContainerComponent>` to also be an `IDistributedApplicationComponentBuilder<IPostgresComponent>`.

Fix #62

eerhardt committed:
Remove ConnectionString on With methods in Hosting App Model

Instead, add new Add overloads that take a connection string to an existing server.

Fix #62

mitchdenny reviewed (approved):
Approved pending comment around manifest generation.

eerhardt committed:
Respond to PR feedback

- Write the connectionString to the manifest if it is provided.
- Use DistributedApplicationException
